### PR TITLE
Fix: #3184 alter miner re-attempt logic to always make second attempt

### LINKED ISSUE #3184: Miner should more aggressively re-attempt block production after first attempt
I've observed this behavior in a mock miner, where the default values

* `first_attempt_time_ms = 5_000`
* `subsequent_attempt_time_ms = 30_000`

Lead to small block mines (i.e., only considering 9 or so transactions).

I *think* what's going on in the mock miner is that the check in `neon_node` for whether or not to attempt another block only attempts a new block if:

1. We've only tried once before (i.e., this is attempt 2) **and** the first attempt was empty.
2. There's new state to mine off of (microblocks or a newly arrived block)

The relevant case branches are roughly

https://github.com/stacks-network/stacks-blockchain/blob/master/testnet/stacks-node/src/neon_node.rs#L1720

What this means is that if the first attempt mines any transactions (even a small amount), and there's no new microblocks being produced (e.g., the last winning miner has microblock production turned off), the miner won't attempt a heavier block ever.

### PR DESCRIPTION
**Pull request by kantai**
This fixes #3184 -- I want to see if this addresses the issue in my mock miner configuration before pushing more on this PR.

### CODE REVIEW

**Review by jcnelson — APPROVED** [review]

**Comment by jcnelson** [comment]
I think this change will make a pretty big difference, but of course I'll wait for the full PR when you've tested it with the mock miner.

**Comment by kantai** [comment]
What are your thoughts on how this should be packaged?

Does this rise to a level of a hotfix, or should I open this against `develop`, and then start a `rc` PR?

**Comment by jcnelson** [comment]
I don't think the chain is at immediate risk right now, so I don't think this rises to the level of hotfix.  This issue causes inconvenience, but nothing more.  I think what you could do is merge this to `develop` and immediately cut a `-rc1`.  We'll just do a release right now with this hotfix applied (and whatever else is in `develop`, which I don't think is very much).  Miners who want to start using this fix can use the `-rc1` artifact, and we'll proceed with a release next week.

File: src/burnchains/bitcoin/spv.rs [hunk 1]
```diff
@@ -229,6 +330,157 @@ impl SpvClient {
...
+            let mut partial = false;
+            let interval_headers = self.read_block_headers(
+                (interval - 1) * BLOCK_DIFFICULTY_CHUNK_SIZE,
+                interval * BLOCK_DIFFICULTY_CHUNK_SIZE,
+            )?;
+            let interval_work = SpvClient::get_interval_work(&interval_headers);
+            work_so_far = work_so_far + interval_work;
+
+            if interval_headers.len() == BLOCK_DIFFICULTY_CHUNK_SIZE as usize {
+                self.store_interval_work(interval - 1, work_so_far)?;
```
Comment (pavitthrap) [inline src/burnchains/bitcoin/spv.rs:429]: When interval = highest_interval + 1, wouldn't you be overwriting the interval work for highest_interval?

File: src/burnchains/burnchain.rs [hunk 2]
```diff
@@ -980,27 +980,24 @@ impl Burnchain {
...
+        let end_block = indexer.sync_headers(highest_header, None)?;
+        let mut start_block = sync_height;
         if db_height < start_block {
             start_block = db_height;
         }
 
         debug!(
             "Sync'ed headers from {} to {}. DB at {}",
-            start_block, end_block, db_height
+            highest_header, end_block, db_height
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/burnchain.rs:1000]: This is an unmerged change from master

File: src/burnchains/bitcoin/indexer.rs [hunk 3]
```diff
@@ -385,47 +387,141 @@ impl BitcoinIndexer {
...
         if start_block > 0 {
-            let start_header = canonical_spv_client
-                .read_block_header(start_block)?
-                .expect(&format!("BUG: missing block header for {}", start_block));
-            reorg_spv_client.insert_block_headers_before(start_block - 1, vec![start_header])?;
+            if start_block > BLOCK_DIFFICULTY_CHUNK_SIZE {
+                if remove_old {
+                    // set up a .reorg db
+                    // * needs the last difficulty interval of headers (note that the current
+                    // interval is `start_block / BLOCK_DIFFICULTY_CHUNK_SIZE - 1).
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:454]: This is an unmerged change from master

File: src/burnchains/bitcoin/indexer.rs [hunk 4]
```diff
@@ -385,47 +387,141 @@ impl BitcoinIndexer {
...
+                    // * needs the last interval's chain work calculation
+                    let interval_start_block =
+                        (start_block / BLOCK_DIFFICULTY_CHUNK_SIZE).saturating_sub(2);
+                    let base_block = interval_start_block * BLOCK_DIFFICULTY_CHUNK_SIZE;
+                    let interval_headers =
+                        canonical_spv_client.read_block_headers(base_block, start_block + 1)?;
+                    assert!(
+                        interval_headers.len() >= (start_block - base_block) as usize,
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:462]: This is an unmerged change from master

File: src/burnchains/bitcoin/indexer.rs [hunk 5]
```diff
@@ -1240,5 +1441,1884 @@ mod test {
...
+            assert!(
+                spv_client.get_headers_height().unwrap() >= 40322,
+                "This test needs headers up to 40320"
+            );
+            spv_client.drop_headers(40320).unwrap();
+        }
+
+        let mut spv_client =
+            SpvClient::new(&db_path, 0, None, BitcoinNetworkType::Mainnet, true, false).unwrap();
+
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:2963]: This is an unmerged change from master

File: src/burnchains/bitcoin/spv.rs (hunk 1 above)
Comment (kantai) [reply src/burnchains/bitcoin/spv.rs:429]: This is an unmerged change from master

File: src/burnchains/bitcoin/indexer.rs [hunk 6]
```diff
@@ -1240,5 +1441,1884 @@ mod test {
...
+                None,
+                BitcoinNetworkType::Mainnet,
+                true,
+                false,
+            )
+            .unwrap();
+
+            assert!(
+                spv_client.get_headers_height().unwrap() >= 40322,
+                "This test needs headers up to 40320"
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:2956]: This is an unmerged change from master

File: src/burnchains/bitcoin/indexer.rs [hunk 7]
```diff
@@ -1240,5 +1441,1884 @@ mod test {
...
+                        .insert_block_headers_after(start_block - 1, ret)
+                        .unwrap();
+                    Ok(())
+                },
+            )
+            .unwrap();
+
+        assert!(inserted_bad_header);
+
+        // reorg is ignored
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:3063]: Yes, that’s the idea, but this is also part of the unmerged-into-develop changes in master (due to a hotfix).

File: testnet/stacks-node/src/neon_node.rs [hunk 8]
```diff
@@ -1717,20 +1725,12 @@ impl StacksNode {
...
-                        best_attempt = 1;
-                        break;
-                    } else if prev_block.attempt == 1 {
-                        // Don't let the fact that we've built an empty block during this sortition
-                        // prevent us from trying again.
-                        best_attempt = 1;
-                        continue;
-                    }
+
+                if prev_block.anchored_block.txs.len() == 1 && prev_block.attempt == 1 {
```
[the remark this answers is not in the document]
Comment (kantai) [reply testnet/stacks-node/src/neon_node.rs:1729]: Yes, this branch is reachable, because here we iterate over each attempt to figure out which attempt was the “best attempt”, so when the last miner blocks length is greater than one, we still end up reading attempt 1 during the iteration.

File: src/burnchains/bitcoin/indexer.rs [hunk 9]
```diff
@@ -385,47 +387,141 @@ impl BitcoinIndexer {
...
+        )
+    }
+
     /// Create a SPV client for starting reorg processing
     fn setup_reorg_headers(
         &mut self,
         canonical_spv_client: &SpvClient,
         reorg_headers_path: &str,
         start_block: u64,
+        remove_old: bool,
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/burnchains/bitcoin/indexer.rs:430]: This is an unmerged change from master (due to a hotfix) — if we want to add documentation to the burn module, we should do that in a separate PR.

**Comment by kantai** [comment]
> Is there any way to see before/after performance for this PR before launching?

A mock miner could provide logging data on the change in behavior. Hiro’s mock miner has been running with this branch, and indeed, this change leads to a second block attempt in instances where it would not have before.

**Comment by gregorycoppola** [comment]
Overall, I think it makes sense to address this. Thanks for the developing/testing the fix.

**Review by gregorycoppola — COMMENTED** [review]
**Is this an Emergency?**

I think we are now treating this PR/issue as a non-emergency. I think this makes sense because we have known about miner sub-optimalities since early Q1, at least. It is not a surprise.

**Norm for Including Data with PR's that Change Performance**

It is a dominant norm in the industry that, if you have a PR that claims to change performance, you will demonstrate that improvement using data, before it gets launched, so you are not testing in production.

In general, this norm would greatly benefit Stacks.

This particular PR is a great example of why the norm is important.

If this PR is claiming to increase throughput. We should know:

* by how much does this change improve throughput?

And we should know this:

* without testing in production

If this PR *does not* claim to improve throughput, we should be clear on that with ourselves and with users, because this PR/issue has been pointed to as a resolution to throughput complaints.

**Conclusion**
If you just want a rubber stamp, I can approve. But, for a non-trivial review, my comments are around the lack of performance testing and data.

Also, I would do the merge of `master->develop` in a separate step, to make it easier for reviewers, and all of us, to understand the change we are making with more confidence and less stress.

File: testnet/stacks-node/src/neon_node.rs [hunk 10]
```diff
@@ -1702,7 +1702,15 @@ impl StacksNode {
         };
 
         // has the tip changed from our previously-mined block for this epoch?
-        let attempt = {
+        let attempt = if last_mined_blocks.len() <= 1 {
```
[the remark this answers is not in the document]
Comment (kantai) [reply testnet/stacks-node/src/neon_node.rs:1705]: Only the first interrupt is "different" from the second interrupt. During the first interrupt, the block assembly is limited to `first_attempt_time_ms` (default 5 seconds). Every attempt thereafter is limited to `subsequent_attempt_time_ms` (default 30 seconds).

The reason to only force the second attempt is that a third or fourth attempt is only "different" if the underlying chainstate is different (which is checked by the else branch) -- they use the same block assembly times, so a heavier block wouldn't necessarily be produced if third or fourth attempts were made.

**Comment by kantai** [comment]
> by how much does this change improve throughput?

Do you have a suggestion for how to measure such a thing? Throughput is an emergent network property which depends on (amongst other things) bitcoin block times, the particular transactions in the mempool at any given time, how many other miners are operating (and what their configurations are).

This particular PR fixes a corner case that is only triggered if microblocks aren't being produced. This is highly dependent on miner configuration. When this PR was opened, few of the active miners were producing microblocks. Currently, most miners _are_ producing microblocks, so this corner case is unlikely to be reached.

Perhaps, because this is a fix for the intended miner behavior, this should just be tested via an integration test?

**Comment by gregorycoppola** [comment]
@kantai 

Sorry for the late reply. I was waiting for the "re-view" signal, but perhaps should have just monitored this thread.

Specific to this PR, I would create an experiment in which [case 1] is the behavior of the system at `develop`, and [case 2] has your changes, and run run a mock miner for each for X hours (e.g., 24 hours).

Then, report the number of transactions in each case, and analyze some statistics (e.g., [case 2] is 5% faster).

One problem with getting performance data might be that there could be different miner settings. The ideal answer to this would be to sample enough settings that we are confident in our picture. At the least, I think we should be able to demonstrate a benefit in *some* parameter setting.

**Comment by gregorycoppola** [comment]
Further to this PR in particular, but also to the general process...

I was looking at follower data and was wondering about the following experiment:
* let each miner finish each block on the first attempt

What would be the throughput in this case? How long would the block take? How many blocks would get missed?

Ideally, we could get data on the performance of both of these hypotheses, and then look at it, and then maybe seeing the data would inspire a third idea, etc.

**Review by gregorycoppola — APPROVED** [review]

**Comment by gregorycoppola** [comment]
I approved in case you guys really want to submit.

I am running mock miner experiments at the moment, and will include the data in a future issue/PR.

All you have to do, is, for `develop` and your PR:
* start the miner
* capture the logs
* run some statistics on the logs

**Comment by kantai** [comment]
I added an integration test to cover the fixed behavior (`tests::neon_integrations::miner_submit_twice`).

Will merge once the CI passes (CI already passed before an update from `develop`: just needed to merge changes to the CHANGELOG).
